### DISCUSSION TITLE
GPU utilization

### DISCUSSION THREAD

**Discussion by NikoYura** [slimgroup/JUDI.jl#271] [Q&A] [answered]
Hello JUDI developers,

I’m encountering an issue where the built-in JUDI tests show GPU utilization correctly after setting the environment variables in my notebook, but when I try to run my own tasks such as forward modeling or Reverse Time Migration (RTM), only the CPU is being utilized. Could you help me find the potential cause of this discrepancy?

Julia Version 1.10.5
JUDI v3.4.6

I would appreciate guidance on the paths I should investigate or potential misconfigurations that could be leading to this behavior.

Thank you in advance for your help!

**Comment by mloubout** [comment]
How do you monitor the gpuy usage, and what's the problem size you are running?

**Reply by NikoYura** [reply]
Usually I have messages that u(x,t) is assuming to fit the GPU memory (I get a lot of them during JUDI/tests). Also even for a small 2D problem I notice some GPU utilization by using nvidia-smi. When I increase the size of a problem in tests functions I see increased utilization. 

I would be grateful if you can suggest other monitors too.

**Reply by mloubout** [reply]
I'd recommend to install nvtop for monitoring GPU utilization

>is assuming to fit the GPU memory

For gradients, that requires saving the full time history yes it will say that, for forward modeling it will not.

**Reply by NikoYura** [reply]
Thank you for your suggestion!

During the tests, memory consuming processes appear (screenshot attached), with the CPU behaving inactive. But during my tasks (forward modeling, RTM with adjoint counting), there are no processes in nvidia-smi, and the CPU is fully loaded.
![pastedImage](https://github.com/user-attachments/assets/99028861-9a32-44db-97eb-c904c4be9c71)

**Reply by mloubout** [reply]
Can you run your own scripts/... with `DEVITO_LOGGING=DEBUG` and show the log?

**Reply by NikoYura** [reply]
Thanks for the tip, I found the problem with the wrong call thanks to debug. A little bit not how I expected the initial behavior of my program.

I would like to clarify, does splitting the algorithm into workers with `distributed.jl` package work similarly with GPU? Are there any nuances?
Is it possible to set environment parameters in the notebook instead of exporting? For some reason some variables work, but GPU variables or for example debug do not work in this form `ENV[“DEVITO_LOGGING”] = “DEBUG”`?

**Reply by mloubout** [reply, marked as answer]
>Is it possible to set environment parameters in the notebook instead of exporting?

There is a utility function `set_devito_config` that allows you to modify the devito configuration from a script or notebook. For example you can do `set_devito_config("log-level", "DEBUG")`

>does splitting the algorithm into workers with distributed.jl package work similarly with GPU? 

Yes it does, you might need to look at the CUDA.jl documentation to make sure each worker is pinned to its own GPU.
